### PR TITLE
MissingRestorePropertyError

### LINKED ISSUE #7218: MissingRestorePropertyError doesn't exist
### Issue Description

Pretty minor issue, but MissingRestorePropertyError isn't a thing, so we shouldn't pretend we can throw it. We should make it a thing.

### Steps to Reproduce

`dotnet build /rp:` on anything.

### Expected Behavior

Some nice error message about how /rp is missing a parameter.

### Actual Behavior

```
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
Microsoft.Build.Shared.InternalErrorException: MSB0001: Internal MSBuild Error: Missing resource 'MissingRestorePropertyError'
   at Microsoft.Build.Shared.ErrorUtilities.ThrowInternalError(String message, Object[] args)
   at Microsoft.Build.Shared.ErrorUtilities.VerifyThrow(Boolean condition, String unformattedMessage, Object arg0)
   at Microsoft.Build.Shared.AssemblyResources.GetString(String name)
   at Microsoft.Build.Shared.ResourceUtilities.FormatResourceStringStripCodeAndKeyword(String& code, String& helpKeyword, String resourceName, Object[] args)
   at Microsoft.Build.Shared.ResourceUtilities.FormatResourceStringStripCodeAndKeyword(String resourceName, Object[] args)
   at Microsoft.Build.CommandLine.CommandLineSwitchException.Throw(String messageResourceName, String commandLineArg, String[] messageArgs)
   at Microsoft.Build.CommandLine.CommandLineSwitchException.Throw(String messageResourceName, String commandLineArg)
   at Microsoft.Build.CommandLine.CommandLineSwitches.ThrowErrors()
   at Microsoft.Build.CommandLine.MSBuildApp.ProcessCommandLineSwitches(CommandLineSwitches switchesFromAutoResponseFile, CommandLineSwitches switchesNotFromAutoResponseFile, String& projectFile, String[]& targets, String& toolsVersion, Dictionary`2& globalProperties, ILogger[]& loggers, LoggerVerbosity& verbosity, List`1& distributedLoggerRecords, Int32& cpuCount, Boolean& enableNodeReuse, TextWriter& preprocessWriter, TextWriter& targetsWriter, Boolean& detailedSummary, ISet`1& warningsAsErrors, ISet`1& warningsAsMessages, Boolean& enableRestore, Boolean& interactive, ProfilerLogger& profilerLogger, Boolean& enableProfiler, Dictionary`2& restoreProperties, Boolean& isolateProjects, GraphBuildOptions& graphBuild, String[]& inputResultsCaches, String& outputResultsCache, Boolean& lowPriority, Boolean recursing)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)
Unhandled exception: Microsoft.Build.Shared.InternalErrorException: MSB0001: Internal MSBuild Error: Missing resource 'MissingRestorePropertyError'
   at Microsoft.Build.Shared.ErrorUtilities.ThrowInternalError(String message, Object[] args)
   at Microsoft.Build.Shared.ErrorUtilities.VerifyThrow(Boolean condition, String unformattedMessage, Object arg0)
   at Microsoft.Build.Shared.AssemblyResources.GetString(String name)
   at Microsoft.Build.Shared.ResourceUtilities.FormatResourceStringStripCodeAndKeyword(String& code, String& helpKeyword, String resourceName, Object[] args)
   at Microsoft.Build.Shared.ResourceUtilities.FormatResourceStringStripCodeAndKeyword(String resourceName, Object[] args)
   at Microsoft.Build.CommandLine.CommandLineSwitchException.Throw(String messageResourceName, String commandLineArg, String[] messageArgs)
   at Microsoft.Build.CommandLine.CommandLineSwitchException.Throw(String messageResourceName, String commandLineArg)
   at Microsoft.Build.CommandLine.CommandLineSwitches.ThrowErrors()
   at Microsoft.Build.CommandLine.MSBuildApp.ProcessCommandLineSwitches(CommandLineSwitches switchesFromAutoResponseFile, CommandLineSwitches switchesNotFromAutoResponseFile, String& projectFile, String[]& targets, String& toolsVersion, Dictionary`2& globalProperties, ILogger[]& loggers, LoggerVerbosity& verbosity, List`1& distributedLoggerRecords, Int32& cpuCount, Boolean& enableNodeReuse, TextWriter& preprocessWriter, TextWriter& targetsWriter, Boolean& detailedSummary, ISet`1& warningsAsErrors, ISet`1& warningsAsMessages, Boolean& enableRestore, Boolean& interactive, ProfilerLogger& profilerLogger, Boolean& enableProfiler, Dictionary`2& restoreProperties, Boolean& isolateProjects, GraphBuildOptions& graphBuild, String[]& inputResultsCaches, String& outputResultsCache, Boolean& lowPriority, Boolean recursing)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String[] commandLine)
   at Microsoft.Build.CommandLine.MSBuildApp.Main(String[] args)
   at Microsoft.DotNet.Cli.Utils.MSBuildForwardingAppWithoutLogging.ExecuteInProc(String[] arguments)
```

### Analysis

We probably just need to add text for the MissingRestorePropertyError, and everything will be good.

### PR DESCRIPTION
**Pull request by jrdodds**
Fixes #7218

### Context
MissingRestorePropertyError doesn't exist. The /rp switch accepts properties to be used during a Restore target. The /p switch accepts properties. The InvalidPropertyError is used by both switches. It appears the intent may have been to also share the MissingPropertyError.

### Changes Made
Edit for ParameterizedSwitch.RestoreProperty to use MissingPropertyError in place of MissingRestorePropertyError.
Edit to unit test PropertySwitchIdentificationTests to also test the ParameterizedSwitch.RestoreProperty.

### Testing
Tested locally built msbuild.exe. An exception is no longer thrown.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7280) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7280) before we can accept your contribution.<br/><br/>:x: jrdodds [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7280)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=7280) it.</sub>

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7280) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7280) before we can accept your contribution.<br/><br/>:x: jrdodds [sign now](https://cla.dotnetfoundation.org/dotnet/msbuild?pullRequest=7280)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/msbuild?pullRequest=7280) it.</sub>

**Review by Forgind — APPROVED** [review]

File: src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs [hunk 1]
```diff
@@ -295,6 +295,29 @@ public void PropertySwitchIdentificationTests(string property)
...
+        public void RestorePropertySwitchIdentificationTests(string property)
+        {
+            CommandLineSwitches.ParameterizedSwitch parameterizedSwitch;
+            string duplicateSwitchErrorMessage;
+            bool multipleParametersAllowed;
+            string missingParametersErrorMessage;
+            bool unquoteParameters;
+            bool emptyParametersAllowed;
+
+            CommandLineSwitches.IsParameterizedSwitch(property, out parameterizedSwitch, out duplicateSwitchErrorMessage, out multipleParametersAllowed, out missingParametersErrorMessage, out unquoteParameters, out emptyParametersAllowed).ShouldBeTrue();
```
Comment (Forgind) [inline src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs:313]:
```suggestion
                CommandLineSwitches.IsParameterizedSwitch(property, out CommandLineSwitches.ParameterizedSwitch parameterizedSwitch, out string duplicateSwitchErrorMessage, out bool multipleParametersAllowed, out string missingParametersErrorMessage, out bool unquoteParameters, out bool emptyParametersAllowed).ShouldBeTrue();
```

Little trick to inline these 🙂

File: src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs [hunk 2]
```diff
@@ -295,6 +295,29 @@ public void PropertySwitchIdentificationTests(string property)
...
+            parameterizedSwitch.ShouldBe(CommandLineSwitches.ParameterizedSwitch.RestoreProperty);
+            duplicateSwitchErrorMessage.ShouldBeNull();
+            multipleParametersAllowed.ShouldBeTrue();
+            missingParametersErrorMessage.ShouldNotBeNull();
```
Comment (Forgind) [inline src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs:317]: Maybe specify that this should be the property error message? You would have to get it out of strings with ResourceUtilities.FormatResourceString or something to ensure it works on any locale.

Comment (jrdodds) [reply src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs:310]: There is a set of these unit tests which are all effectively testing the `CommandLineSwitches.IsParameterizedSwitch()` method. That method doesn't retrieve the resource string. It pulls a `ParameterizedSwitchInfo` object from the `s_parameterizedSwitchesMap` array.

The `RestorePropertySwitchIdentificationTests` test could be modified to check that the value of `missingParametersErrorMessage` is `"MissingPropertyError"`.

Should all of the unit tests with `missingParametersErrorMessage.ShouldNotBeNull();` be updated accordingly? (I'm happy to make the changes but I don't want to presume.)

Comment (Forgind) [reply src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs:310]: That sounds good to me. As far as changing any of the other tests, it sounds reasonable to me, but it's less relevant to the main point of this PR, so if you decide to change them all, I'd put that in a separate PR.

Comment (jrdodds) [reply src/MSBuild.UnitTests/CommandLineSwitches_Tests.cs]: The unit test has been updated. Thanks

**Comment by Forgind** [comment]
Copied from issue:
### Examples
#### Property switch missing property
```
msbuild test.proj /p
```
generates error
```
MSBUILD : error MSB1005: Specify a property and its value.
Switch: /p
```

#### RestoreProperty switch missing property
```
msbuild test.proj /rp
```
generates error
```
MSBUILD : error MSB1005: Specify a property and its value.
Switch: /rp
```

#### RestoreProperty and Property switches both missing property
```
msbuild test.proj /rp /p
```
generates error
```
MSBUILD : error MSB1005: Specify a property and its value.
Switch: /rp
```
If the command line has multiple errors, only one error is reported.

**Comment by rainersigwald** [comment]
> ```
> msbuild test.proj /rp
> ```
> 
> generates error
> 
> ```
> MSBUILD : error MSB1005: Specify a property and its value.
> Switch: /rp
> ```

This looks fine since it still makes clear that it's the `/rp` switch that is in error.

**Review by rainersigwald — APPROVED** [review]

**Pull request merged**